nsqd/nsqd.go

```diff
@@ -263,7 +265,9 @@ func (n *NSQD) Main() {
 
 	tcpServer := &tcpServer{ctx: ctx}
 	n.waitGroup.Wrap(func() {
-		protocol.TCPServer(n.tcpListener, tcpServer, n.logf)
+		if err := protocol.TCPServer(n.tcpListener, tcpServer, n.logf); err != nil {
+			go n.Exit()
```

Comment:
I suggest checking the exiting flag here, and if it is not already set, then go Exit().

The flag variable could be an atomic int, to avoid the need to take a lock.

Reply:
(this suggestion would require Exit() to set the flag as the first thing it does)

Reply:
I suggest we keep the check logic be within the `Exit` function. With this we should not take care about the internal exit status flag check before `Exit` is called in other places. And we should keep `Exit` function to be called only once.

With setting and checking separated with aotmic values, race condition may happens. For example, a normal exit and accept error exit happens simultaneously, before the normal exit call set the Exit flag, accept error exit checks it and then call Exit again. Although nsqd will exit correctly but it is called more than once.

With lock like this. Exit function logic except closing listeners will only be called once.

Reply:
BTW, i am not quite sure about your intention, lock consumption should be ok and will not invoke any new performance problems, because all this happens under exit phase.

Reply:
The reason I suggested the exiting flag early on is because deciding to call Exit here could be as simple as:

```diff
--- a/nsqd/nsqd.go
+++ b/nsqd/nsqd.go
@@ -71,6 +71,7 @@ type NSQD struct {
 
        notifyChan           chan interface{}
        optsNotificationChan chan struct{}
+       exitFlag             int32
        exitChan             chan int
        waitGroup            util.WaitGroupWrapper
 
@@ -264,6 +265,10 @@ func (n *NSQD) Main() {
        tcpServer := &tcpServer{ctx: ctx}
        n.waitGroup.Wrap(func() {
                protocol.TCPServer(n.tcpListener, tcpServer, n.logf)
+               if atomic.LoadInt32(&t.exitFlag) == 0 {
+                       // abnormal listen loop exit
+                       go n.Exit()
+               }
        })
        httpServer := newHTTPServer(ctx, false, n.getOpts().TLSRequired == TLSRequired)
        n.waitGroup.Wrap(func() {
@@ -423,6 +428,10 @@ func (n *NSQD) PersistMetadata() error {
 }
 
 func (n *NSQD) Exit() {
+       if !atomic.CompareAndSwapInt32(&t.exitFlag, 0, 1) {
+               return
+       }
+
        if n.tcpListener != nil {
                n.tcpListener.Close()
        }
```

(but there still is the issue of exiting with an error code)

Reply:
Agreed that the atomic would be slightly cleaner, but I prefer directing flow through the existing machinery that's setup to manage the lifecycle of the service rather than introducing a new exceptional path — to ensure that, whatever `svc.Run` is doing, it gets done.

Reply:
yeah ... calling NSQD.Exit() cleans up but does not actually exit ... go-svc needs to be involved anyway

Reply:
one more cheesy idea to avoid needing to deal with go-svc:

```diff
--- a/nsqd/nsqd.go
+++ b/nsqd/nsqd.go
@@ -264,6 +264,12 @@ func (n *NSQD) Main() {
        tcpServer := &tcpServer{ctx: ctx}
        n.waitGroup.Wrap(func() {
                protocol.TCPServer(n.tcpListener, tcpServer, n.logf)
+               if atomic.LoadInt32(&t.exitFlag) == 0 {
+                       go func() {
+                               n.Exit()
+                               os.Exit(1)
+                       }()
+               }
        })
```

Reply:
Meh, let's just do it the "right" way.